executor/load_data.go

```diff
@@ -461,32 +554,54 @@ func (e *LoadDataInfo) SetMessage() {
 }
 
 func (e *LoadDataInfo) colsToRow(ctx context.Context, cols []field) []types.Datum {
-	totalCols := e.Table.Cols()
-	for i := 0; i < len(e.row); i++ {
-		if i >= len(cols) {
+	row := make([]types.Datum, 0, len(e.insertColumns))
+
+	for i := 0; i < len(e.FieldMappings); i++ {
+		if i >= len(cols) && e.FieldMappings[i].Column != nil {
 			// If some columns is missing and their type is time and has not null flag, they should be set as current time.
-			if types.IsTypeTime(totalCols[i].Tp) && mysql.HasNotNullFlag(totalCols[i].Flag) {
-				e.row[i].SetMysqlTime(types.CurrentTime(totalCols[i].Tp))
+			if types.IsTypeTime(e.FieldMappings[i].Column.Tp) && mysql.HasNotNullFlag(e.FieldMappings[i].Column.Flag) {
+				row = append(row, types.NewTimeDatum(types.CurrentTime(e.FieldMappings[i].Column.Tp)))
 				continue
 			}
-			e.row[i].SetNull()
+			row = append(row, types.NewDatum(nil))
 			continue
 		}
+
+		var val types.Datum
 		// The field with only "\N" in it is handled as NULL in the csv file.
 		// See http://dev.mysql.com/doc/refman/5.7/en/load-data.html
 		if cols[i].maybeNull && string(cols[i].str) == "N" {
-			e.row[i].SetNull()
+			val = types.NewDatum(nil)
```

Comment:
Is it possible that if condition result is like `i >= len(cols) && e.FieldMappings[i].Column == nil` and using `i` to visit `cols` or `ColumnsAndUserVars` is not safe? Maybe we could do some refactor to make `colsToRow` function more clear and easy to maintain, like 
```
for each column 
    if the value exists in `csv` data
    else if the value from user variable calculation
    else use the default value or null value
return this new row
```

Reply:
Thanks for your comment!
It's unsafe to access elements in `ColumnsAndUserVars` here, i fixed it and refactored input values preprocessing inside the for loop.